components/brave_wallet/browser/internal/orchard_storage/orchard_types.h

```diff
@@ -0,0 +1,116 @@
+  OrchardCheckpoint();
+  OrchardCheckpoint(CheckpointTreeState, std::vector<uint32_t>);
+  ~OrchardCheckpoint();
+  OrchardCheckpoint(const OrchardCheckpoint& other);
+  OrchardCheckpoint& operator=(const OrchardCheckpoint& other);
+  OrchardCheckpoint(OrchardCheckpoint&& other);
+  OrchardCheckpoint& operator=(OrchardCheckpoint&& other);
+
+  bool operator==(const OrchardCheckpoint& other) const = default;
+
+  // Leaf position of the checkpoint.
+  CheckpointTreeState tree_state_position;
+  // List of note positions that were spent at this checkpoint.
+  std::vector<uint32_t> marks_removed;
+};
+
+struct OrchardCheckpointBundle {
+  OrchardCheckpointBundle(uint32_t checkpoint_id, OrchardCheckpoint);
+  ~OrchardCheckpointBundle();
+  OrchardCheckpointBundle(const OrchardCheckpointBundle& other);
+  OrchardCheckpointBundle& operator=(const OrchardCheckpointBundle& other);
+  OrchardCheckpointBundle(OrchardCheckpointBundle&& other);
+  OrchardCheckpointBundle& operator=(OrchardCheckpointBundle&& other);
+
+  bool operator==(const OrchardCheckpointBundle& other) const = default;
+
+  // The block height serves as the checkpoint identifier.
+  uint32_t checkpoint_id = 0;
+  OrchardCheckpoint checkpoint;
+};
+
+// Address of a subtree in the shard tree.
+struct OrchardShardAddress {
+  uint8_t level = 0;
+  uint32_t index = 0;
+
+  bool operator==(const OrchardShardAddress& other) const = default;
+};
+
+// Top part of the shard tree from the root to the shard roots level
+// Used for optimization purposes in the shard tree crate.
+using OrchardShardTreeCap = std::vector<uint8_t>;
+
+// Subtree with root selected from the shard roots level.
+struct OrchardShard {
+  OrchardShard();
+  OrchardShard(OrchardShardAddress shard_addr,
+               std::optional<OrchardShardRootHash> shard_hash,
+               std::vector<uint8_t> shard_data);
+  ~OrchardShard();
+
+  OrchardShard(const OrchardShard& other);
+  OrchardShard& operator=(const OrchardShard& other);
+  OrchardShard(OrchardShard&& other);
+  OrchardShard& operator=(OrchardShard&& other);
+
+  bool operator==(const OrchardShard& other) const = default;
+
+  // Subtree root address.
+  OrchardShardAddress address;
+  // Root hash exists only on completed shards.
+  std::optional<OrchardShardRootHash> root_hash;
+  std::vector<uint8_t> shard_data;
+  // Right-most position of the subtree leaf.
+  uint32_t subtree_end_height = 0;
+};
+
+struct OrchardCommitment {
+  OrchardCommitmentValue cmu;
+  bool is_marked = false;
+  std::optional<uint32_t> checkpoint_id;
+};
+
+// Compact representation of the Merkle tree on some point.
+// Since batch inserting may contain gaps between scan ranges we insert
+// frontier which allows to calculate node hashes and witnesses(merkle path from
+// leaf to the tree root) even when previous scan ranges are not completed.
+struct OrchardTreeState {
+  OrchardTreeState();
+  ~OrchardTreeState();
+  OrchardTreeState(const OrchardTreeState&);
```

Comment:
Based on a comment I left regarding the use of this struct above, I think it would be good to review all the structs we have in this file to determine which of them do in fact need to support copy, and which ones would benefit from being move-only structs, so we avoid unnecessary copies of these vectors.

Of course, if copy is a sensible thing for some of these types we should just delete the copy assignment/constructor, but it is preferable to only support move, if we can.

If you need some reference of how to make a type move-only you can take a look at the style guide https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types.

Reply:
Fixed